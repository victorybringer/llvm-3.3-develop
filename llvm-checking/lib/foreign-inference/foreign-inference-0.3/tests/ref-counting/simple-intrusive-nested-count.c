#include <stdlib.h>

typedef struct {
  int rc;
} refcount_t;

typedef struct {
  refcount_t ref_count;
} cairo_surface_t;

cairo_surface_t* cairo_surface_create() {
  cairo_surface_t *ret = malloc(sizeof(cairo_surface_t));
  ret->ref_count.rc = 1;

  return ret;
}

cairo_surface_t* cairo_surface_reference(cairo_surface_t *s) {
  ++s->ref_count.rc;

  return s;
}

void cairo_surface_destroy(cairo_surface_t *s) {
  if(--s->ref_count.rc) {
    return;
  }

  free(s);
}
