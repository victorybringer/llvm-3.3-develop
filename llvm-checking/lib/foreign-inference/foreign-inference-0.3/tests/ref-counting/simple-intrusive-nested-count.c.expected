fromList [ ("cairo_surface_destroy", "cairo_surface_reference") ]
